File: .old/Aclass_Python/A1126.py
num_vertice, num_edge = list(map(int, input().split()))
edges = [[] for _ in range(num_vertice + 1)]
for _ in range(num_edge):
    a, b = list(map(int, input().split()))
    edges[a].append(b)
    edges[b].append(a)

degree = [len(edges[i]) for i in range(num_vertice + 1)]
odd = len([i for i in degree if i % 2 == 1])

#dfs
count = 0
visited = [False for i in range(num_vertice + 1)]
visited[1] = True
def dfs(i):
    global count
    count += 1
    for neighbor in edges[i]:
        if not visited[neighbor]:
            visited[neighbor] = True
            dfs(neighbor)

print(" ".join(list(map(str, degree[1:]))))
if odd == 0:
    dfs(1)
    if count == num_vertice:
        print("Eulerian")
    else:
        print("Non-Eulerian")
elif odd == 2:
    dfs(1)
    if count == num_vertice:
        print("Semi-Eulerian")
    else:
        print("Non-Eulerian")
else:
    print("Non-Eulerian")

File: .old/Aclass_one_more_time/A1126.cpp
#include <iostream>
#include <vector>
using namespace std;
vector<int> graph[501];
int count = 0;
bool visited[501] = {false};
void dfs(int index){
    count++;
    for(int i = 0; i < graph[index].size(); i++){
        if(!visited[graph[index][i]]) {
            visited[graph[index][i]] = true;
            dfs(graph[index][i]);
        }
    }
}

int main(){
    int num_vertice, num_edge;
    bool first = true;
    int odd = 0;
    cin >> num_vertice >> num_edge;
    for(int i = 0; i < num_edge; i++){
        int v1, v2;
        cin >> v1 >> v2;
        graph[v1].push_back(v2);
        graph[v2].push_back(v1);
    }
    for(int i = 1; i <= num_vertice; i++){
        if(first) {
            cout << graph[i].size();
            first = false;
        } else cout << " " << graph[i].size();
        if(graph[i].size() % 2 != 0) odd++;
    }
    cout << endl;
    visited[1] = true;
    if(odd == 0) {
        dfs(1);
        if(count == num_vertice) cout << "Eulerian" << endl;
        else cout << "Non-Eulerian" << endl;
    } else if(odd == 2){
        dfs(1);
        if(count == num_vertice) cout << "Semi-Eulerian" << endl;
        else cout << "Non-Eulerian" << endl;
    } else cout << "Non-Eulerian" << endl;
    return 0;
}
